### PR TITLE
#218 enable cross fork compare

### PR DESCRIPTION
**Pull request by if6was9**
this addresses #218

### CODE REVIEW

File: src/main/java/org/kohsuke/github/GHRepository.java [hunk 1]
```diff
@@ -682,7 +682,25 @@ public GHCompare getCompare(GHCommit id1, GHCommit id2) throws IOException {
...
+
+        try {
+            // try to grab the org name and qualify the branch name
+            String owner1 = id1.getOwner().getOwnerName();
+            String owner2 = id2.getOwner().getOwnerName();
+            if (!owner1.equals(owner2)) {
+                fqid1 = String.format("%s:%s",owner1,id1.getName());
+                fqid2 = String.format("%s:%s",owner2,id2.getName());
+            }
+        }
```
Comment (lanwen) [inline src/main/java/org/kohsuke/github/GHRepository.java]: catch should be after } -> `} catch {`

File: src/main/java/org/kohsuke/github/GHRepository.java [hunk 2]
```diff
@@ -682,7 +682,25 @@ public GHCompare getCompare(GHCommit id1, GHCommit id2) throws IOException {
...
+        catch (RuntimeException e) {
```
Comment (lanwen) [inline src/main/java/org/kohsuke/github/GHRepository.java]: bad practice to catch everything and ignore. Should catch only when we request owner.

File: src/main/java/org/kohsuke/github/GHRepository.java [hunk 3]
```diff
@@ -682,7 +682,25 @@ public GHCompare getCompare(GHCommit id1, GHCommit id2) throws IOException {
...
-        return getCompare(id1.getName(),id2.getName());
+
+        String fqid1 = id1.getName();
+        String fqid2 = id2.getName();
+
+        try {
+            // try to grab the org name and qualify the branch name
+            String owner1 = id1.getOwner().getOwnerName();
+            String owner2 = id2.getOwner().getOwnerName();
+            if (!owner1.equals(owner2)) {
```
Comment (lanwen) [inline src/main/java/org/kohsuke/github/GHRepository.java]: use StringUtils.equals(s1, s2) as NPE-safe version

**Comment by lanwen** [comment]
:-1: For now.

I think it should be:
1. Get the owners. (Can it be exception here? If not, skip 2 step)
2. Use default owner names if exc.
3. If owners equals, use `getCompare(id1.getName(), id2.getName());`
4. Else format owner:branch and use `getCompare` with that

File: src/main/java/org/kohsuke/github/GHRepository.java [hunk 4]
```diff
@@ -682,7 +682,24 @@ public GHCompare getCompare(GHCommit id1, GHCommit id2) throws IOException {
     }
 
     public GHCompare getCompare(GHBranch id1, GHBranch id2) throws IOException {
+
```
Comment (lanwen) [inline src/main/java/org/kohsuke/github/GHRepository.java]: please avoid empty lines in count more than one

File: src/main/java/org/kohsuke/github/GHRepository.java [hunk 5]
```diff
@@ -682,7 +682,23 @@ public GHCompare getCompare(GHCommit id1, GHCommit id2) throws IOException {
...
+        GHRepository owner1 = id1.getOwner();
+        GHRepository owner2 = id2.getOwner();
+
+        // If the owner of the branches is different, we have a cross-fork compare.
+        if (owner1!=null && owner2!=null) {
+            String ownerName1 = owner1.getOwnerName();
+            String ownerName2 = owner2.getOwnerName();
+            if (!StringUtils.equals(ownerName1, ownerName2)) {
+                String qualifiedName1 = String.format("%s:%s",ownerName1,id1.getName());
```
Comment (lanwen) [inline src/main/java/org/kohsuke/github/GHRepository.java]: please add space after `,`

File: src/main/java/org/kohsuke/github/GHRepository.java [hunk 6]
```diff
@@ -682,7 +682,23 @@ public GHCompare getCompare(GHCommit id1, GHCommit id2) throws IOException {
...
+                String qualifiedName2 = String.format("%s:%s",ownerName2,id2.getName());
+                return getCompare(qualifiedName1,qualifiedName2);
```
Comment (lanwen) [inline src/main/java/org/kohsuke/github/GHRepository.java]: and here

**Comment by lanwen** [comment]
:+1: after some code style fixes

**Comment by oleg-nenashev** [comment]
As I see from the code, null owners are not supposed to happen. BTW I'm not sure I've investigated all potential cases.

:+1: in any case, because the original behavior smells like a bug

**Pull request merged**